core/hibernate-search/src/main/java/io/thorntail/hibernate_search/impl/QueryFactory.java

```diff
@@ -0,0 +1,48 @@
+package io.thorntail.hibernate_search.impl;
+
+import java.lang.annotation.Annotation;
+import javax.enterprise.inject.Produces;
+import javax.enterprise.inject.spi.InjectionPoint;
+import org.hibernate.search.jpa.FullTextEntityManager;
+import org.hibernate.search.query.dsl.QueryBuilder;
+import io.thorntail.hibernate_search.QueryBuilderContext;
+import javax.enterprise.inject.Disposes;
+import javax.persistence.EntityManagerFactory;
+import javax.persistence.PersistenceUnit;
+import org.hibernate.search.jpa.Search;
+
+/**
+ * @author kg6zvp
+ */
+public class QueryFactory {
+    @PersistenceUnit
+    EntityManagerFactory emf;
+
+    @Produces
+    @QueryBuilderContext(Object.class)
+    public QueryBuilder getQueryBuilder(InjectionPoint ip) {
+        QueryBuilderContext qb = getQueryAnnotation(ip);
+        if (qb == null) {
+            throw new RuntimeException("@QueryBuilderQualifier not present on QueryBuilder InjectionPoint"); //Will be replaced with custom exception and better message if QueryBuilder feature is required
```

Comment:
This is where using `@Inject` on `QueryBuilder` gets messy, as each instance of a `QueryBuilder` pertains to a single Entity class. It was mentioned this would be a handy inclusion, but this seems messy.

Feedback and comments appreciated as always.

Reply:
I'm not certain, but I'm not sure `Object.class` is needed on `@QueryBuilderContext`, or there's some other way to make it generic. @mkouba should know better than me.

Also, If we're producing an instance for `@QueryBuilderContext`, should qb ever be null? If it was wouldn't that imply we're producing an instance for an injection point that doesn't have the annotation?

Reply:
Using Object.class was the only way to let `@QueryBuilderContext` require a `Class` property.

We're producing an instance of `QueryBuilder`. `@QueryBuilder`Context` is an annotation I added because instances of `QueryBuilder` themselves must be built for a specific entity class.

As to `@QueryBuilderContext` being null, I will add an observer for the event that CDI fires as it is scanning fields annotated with `@Inject` if it turns out we actually want `QueryBuilder` injected. The observer would act as a validator to make sure every `QueryBuilder` field annotated with `@Inject` had a `@QueryBuilderContext` annotation on it. I could also validate that the given `Class` is an entity in that persistence unit, but I don't think there's a need since Hibernate Search would throw an error if a `QueryBuilder` was attempted to be built without a valid `@Entity`-annotated class belonging to the default persistence unit.

Reply:
I don't think you need to observe the event or anything like that. My understanding is that by virtue of the `@Producer` being marked with `@QueryBuilderContext` that an injection point that just specifies a return type of `QueryBuilder` without the annotation would not trigger the producer.

Maybe it's worth using a CDI extension to output a warning that an injection for `QueryBuilder` was made without a corresponding `@QueryBuilderContext` so an instance won't be injected? Now realizing that this is maybe what you meant above

Reply:
@kenfinnigan Yes, that is dead code from an earlier implementation.

Yes, CDI `Extension` is exactly what I meant as they are allowed to observe such lifecycle events. Do we have a convention for custom errors that I should follow?

Reply:
Nothing other than them being created in a message class like https://github.com/thorntail/thorntail/blob/4.x/core/jpa-support/src/main/java/io/thorntail/jpa/impl/JpaMessages.java